### PR TITLE
Rework cert renewal process

### PR DESCRIPTION
**Pull request by guillaumerose**
cert renewal is not accurate (at least from what I saw). We only have one cert to approve. The second one is automatically approved by the cluster machine approver. This PR removes the race.

### CODE REVIEW

File: pkg/crc/machine/machine.go
Comment (praveenkumar) [inline pkg/crc/machine/machine.go:398]: This is now not required because we are using `192.168.126.11` as kubelet IP but before 4.4 timeframe when we had different IP for kubelet then it was required to approve pending node certificate.

File: Makefile
Comment (praveenkumar) [inline Makefile:3]: @guillaumerose This change shouldn't be part of this commit message.

File: pkg/crc/cluster/cluster.go
Comment (praveenkumar) [inline pkg/crc/cluster/cluster.go:101]: Why we are adding thee check, this is not something user facing and we automatic add it as part of code, I am not sure what benefit it provide, if it is not set that mean our code which set this is buggy and need to be fixed.

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go:239]: done :+1:

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go:726]: In case the file doesn't exist, an error will happen and the copy code just after will be triggered.

File: Makefile
Comment (guillaumerose) [reply Makefile:3]: done :+1:

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:101]: I would like to remove most of the branches in the `machine.Start` code. If the creation code could be the same as re-start code, it would be easier to follow. 

It also ensures after each start that the clusterID, the pull secret, etc. are correctly in place.

**Comment by praveenkumar** [comment]
Except https://github.com/code-ready/crc/pull/1532#discussion_r491879049, lgtm

**Comment by guillaumerose** [comment]
I refactored a bit the Start() function to reduce its size. Let me know what you think!

**Comment by praveenkumar** [comment]
@guillaumerose conflict need to be resolved, I do like separating the `start` functionality in it's own file.

File: pkg/crc/cluster/cert-renewal.go
Comment (cfergeau) [inline pkg/crc/cluster/cert-renewal.go]: Half of the changes in this commit replace `err = foo(); if err != nil {}` with `if err := foo(); err != nil {}` which I'm not necessarily opposed to, but would probably be better done in a separate commit rather than cluttering this one.

**Review by cfergeau — COMMENTED** [review]
` Be selective when approving CSR. Either it's a client cert or a serve… `
Short log is too long, even github UI split it :) It's imo too terse to explain what this is about.

File: pkg/crc/cluster/cluster.go
Comment (cfergeau) [reply pkg/crc/cluster/cluster.go:101]: this rationale belongs in the commit log

**Review by cfergeau — COMMENTED** [review]
` Simplify CheckCertsValidity: use a boolean` please explain what your rationale is rather than leaving this as an exercise for the readers.
```cert-renewal: Simplify CheckCertsValidity

All that is needed from CheckCertsValidity() is to indicate whether certificates have expired and need to be renewed, or if no further action is needed. This can be done through a boolean rather than using the CertExpiryState enum as done currently.
```

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go:321]: Good opportunity to fix the spacing?
```
<blank line>
logging.Info("Check internal and public DNS query ...")
if queryOutput, err := dns.CheckCRCPublicDNSReachable(...)
```

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go]: slight preference for this as it's easier to parse:
```
machineConfig := config.MachineConfig {
...
}
_, err := createHost(libMachineClient, machineConfig)
```
(nit: it's neither a renaming nor related to `IsRunning` -> this could actually be done in `Extract functions that return large structs`)

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go:290]: `machineConfigNew()`?

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go:304]: `postStartConfigNew` (or `newPostStartConfig`)

**Review by cfergeau — COMMENTED** [review]
> ` Always generate SSH keys if they don't exist`

Please explain in the commit log the rationale for doing this, why this is better than the current (working) code.


> `Use a struct to memoize the the pull secret` -> memorize (store ?) Commit log is missing some details regarding what your intent is.


> `cluster.RegenerateCertificates doesn't need to start kubelet anymore.`

Before this change, `cluster.RegenerateCertificates` was stopping the kubelet service, and it was restarted after it returned. Maybe a `systemctl restart kubelet` will still be needed, not sure how well it copes with cert renewals without restarts these days.

> `Wait for kubelet certs to be renewed`

This moves lots of code around, I'd expect the commit log to give a summary of what changed.

>` Ensure clusterID and pull secret are correctly set`

The commit log should explain why you want to do this. I see a (very good) reason for doing this, but that's just my guess, so better to spell it explicitly in the commit log so that everyone is on the same line.


> ` Move machine.Start in its own file `

As in the previous commits, stating explicitly why you do that (because it's huge I guess) never hurts :)


> ` Extract method that check DNS on the host and in the VM `

Why? (actually here I thought the commit was doing something entirely different from its log...) `start: Add helper for host/VM DNS checks`?


> ` Rename variables, remove machine.IsRunning `

Can you at least keep the "machine.IsRunning" removal separate? You could give the list of renames you are doing in the commit log so that they are easier to discuss if people disagree. As far as I can see, its `crcBundleMetadata` -> `bundle` and `libMachineAPIClient` -> `libMachineClient`, `vmState` -> `vmStatus`


> ` GetOpenshiftVersion() always return a value `

```
Remove handling of GetOpenshiftVersion() returning ""

Bundles shipped in current crc versions always have the OpenShift version set in their metadata, so we can remove the fallback code which was there in case it's not set.
```

> `Extract functions that return large structs`

Once again, not really clear what the commit is about :(

```
Add helpers to create MachineConfig and PostStartConfig

These struct have quite a few fields to initialise, this commit moves their creation to dedicated helpers.
```

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:101]: added

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go]: My issue with this is that the private key of the runner is mutated in a sub-function of machine.Start. It's not obvious when reading start code that the ssh runner will mutate.

Now, starting here, we are sure the private key is `constants.GetPrivateKeyPath()`.

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go]: The commit message is more precise now.

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go]: I ended up the serie with `EnsurePullSecretPresentInTheCluster` and `EnsurePullSecretPresentOnInstanceDisk`. Is it ok?

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:264]: I renamed `pullSec` to `pullSecret`, but kept content since it used 2 lines after.

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:259]: Constant extracted!

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go:321]: done

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go]: IsRunning is gone, removed by an other commit.

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go:290]: renamed !

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go:304]: renamed !

**Comment by guillaumerose** [comment]
I think I went on all messages. It should be good for an other round!

**Review by cfergeau — COMMENTED** [review]
The `SetPrivateKey() is also removed so that sshRunner is not mutated in a sub function of machine.Start().` log is obsolete since you addressed this differently in master

File: pkg/crc/machine/machine.go
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: `test -e /home/core/.ssh/authorized_keys` seems to be doing the job

**Comment by cfergeau** [comment]
There is a `memoize` typo in one of the commit logs

**Comment by guillaumerose** [comment]
All done! For memoize/memorize, I was referring https://en.wikipedia.org/wiki/Memoization

File: pkg/crc/cluster/cluster.go
Comment (cfergeau) [reply pkg/crc/cluster/cluster.go:104]: >  I am not sure what benefit it provide, if it is not set that mean our code which set this is buggy and need to be fixed.

It would help to recover in case there was a transient cluster failure when the cluster was created, or if the user pressed ctrl+c during the initial cluster creation, ...

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go:90]: fwiw, after applying this PR, I still get
```
$ git grep vmState
pkg/crc/machine/machine.go:     vmState, err := host.Driver.GetState()
pkg/crc/machine/machine.go:             State:         vmState,
pkg/crc/machine/start.go:       vmState, err := host.Driver.GetState()
pkg/crc/machine/start.go:               Status:         vmState.String(),
```

File: pkg/crc/machine/start.go
Comment (cfergeau) [inline pkg/crc/machine/start.go:314]: Not the right commit, this belongs to `Add helper for host/VM DNS checks`

**Review by cfergeau — COMMENTED** [review]
This PR seems to be a bit of multiple independant PRs, I'd expect `Always make sure the pull secret is in the cluster` and `Always make sure the cluster ID is not empty` to come before the commits related to certificate regeneration, so that we have first the commits making VM creation/VM start more similar, and then the certificate regeneration commits

**Review by cfergeau — COMMENTED** [review]
One note regarding `Move machine.Start in its own file`, this impacts `tig  blame`, a bit `git log`, ...

`Verify both client and server kubelet certificates` introduces a significant behaviour change in my tests, this seems to be unexpected from my reading of the commit log?

**Comment by cfergeau** [comment]
At this point, I'd be inclined to suggest switching to 2 PRs, one with the cert renewal commits:
```
215cd38a Wait for kubelet certs to be renewed
af2ed7eb Move code related to certs in the same go file
90ea3320 Make K8sResource a private struct
3cf6c3d9 Verify both client and server kubelet certificates
25ca1e47 Move cluster.RegenerateCertificates right after kubelet startup in machine.Start()
```
and the other with the rest of the commits (but the introduction of `start.go` makes this slightly painful :( The other commits look good to me. Since they are not really related to cert renewal, it would be good to land them while taking a closer look at the cert renewal  changes.

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go]: done in https://github.com/code-ready/crc/pull/1548

File: pkg/crc/cluster/cluster.go
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go]: done in https://github.com/code-ready/crc/pull/1548

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go]: done in https://github.com/code-ready/crc/pull/1548

File: pkg/crc/cluster/cert-renewal.go
Comment (guillaumerose) [reply pkg/crc/cluster/cert-renewal.go:45]: Apparently yes, kubelet uses the new certificate without restarting.

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go]: Dropped this commit

File: pkg/crc/machine/start.go
Comment (guillaumerose) [reply pkg/crc/machine/start.go]: I dropped this commit, it should be ok for this one.

**Comment by guillaumerose** [comment]
The log is far more lighter now with the PR split.

**Comment by guillaumerose** [comment]
About the second cert approval, I triggered the race condition yesterday.

Both crc and machine approver want to approve the kubelet serving certificate. If crc wins, it works. If machine approver wins, crc command hangs.

**Comment by guillaumerose** [comment]
PR rebased. PTAL.

**Comment by cfergeau** [comment]
Still unclear whether this PR is expected to trigger significant behaviour change on `crc start`, after re-reading the commit logs, my feeling is that this is unexpected?

**Comment by guillaumerose** [comment]
There is no significant changes, just bug fixes. It changes:
* Previously, we checked client cert expiry date and waited both the client and serving cert to be rotated. The serving cert might not need to be rotated, but crc still wait for a CSR to approve and hangs. This PR removes the dead lock.
* If the serving cert is expired, crc try to approve it. But, an other process, the cluster machine approver, also tries to do it. If crc wins, no locks. If the approver wins, crc is in a dead lock. This PR removes the approval of the serving cert.

**Comment by cfergeau** [comment]
> There is no significant changes, just bug fixes. It changes:
> 
>     * Previously, we checked client cert expiry date and waited both the client and serving cert to be rotated. The serving cert might not need to be rotated, but crc still wait for a CSR to approve and hangs. This PR removes the dead lock.
> 
>     * If the serving cert is expired, crc try to approve it. But, an other process, the cluster machine approver, also tries to do it. If crc wins, no locks. If the approver wins, crc is in a dead lock. This PR removes the approval of the serving cert.

Could we be now trying too hard/doing too much? With the 4.5.9 bundle from 1.16 and git master:
```
$ crc delete
$ crc start
[...]
INFO Verifying validity of the cluster certificates ...
DEBU About to run SSH command:
date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-client-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2020-10-13T10:06:06+00:00
DEBU Creating /var/srv/dnsmasq.conf with permissions 0644 in the CRC VM
DEBU About to run SSH command with hidden output                              
DEBU SSH command succeeded                      
DEBU About to run SSH command:
sudo podman rm -f dnsmasq              
[...]
```
With this PR and the same bundle:
```
$ crc delete
$ crc start
[...]
INFO Verifying validity of the kubelet certificates ...                                                                                                                                               
DEBU About to run SSH command:
date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-client-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2020-10-13T10:06:06+00:00
DEBU About to run SSH command:
date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-server-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2020-09-14T09:06:42+00:00
DEBU Certs have expired, they were valid till: 14 Sep 20 09:06 +0000
[...]
INFO Cluster TLS certificates have expired, renewing them... [will take up to 5 minutes]
```
With git master, if I run `date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-server-current.pem -noout -enddate | cut -d= -f 2)` in the VM right after crc start, the server cert is valid for 1 more week without any explicit action from crc.

**Comment by guillaumerose** [comment]
Yes the serving cert is an extra check. With current master, it is silently updated (during the 3min sleep time maybe). Until the cert is updated, user can't exec in a container nor fetching container logs. 
I hope to remove the final sleep(3 min) with this kind of checks.

File: pkg/crc/machine/machine.go
Comment (cfergeau) [reply pkg/crc/machine/machine.go:412]: I'd add an explanation about this in the commit log. Especially as this seems to be silently removed in ` Wait for kubelet certs to be renewed`

Comment (guillaumerose) [reply pkg/crc/machine/machine.go:412]: Added an extra commit for this!

File: pkg/crc/cluster/cert-renewal.go
Comment (cfergeau) [inline pkg/crc/cluster/cert-renewal.go:38]: It's no longer first/second, but client/server CSRs. I guess it would be possible to only get the second CSR for example in some scenarios.

File: pkg/crc/cluster/cert-renewal.go
Comment (cfergeau) [inline pkg/crc/cluster/cert-renewal.go:52]: Have you considered adding a `waitForCertRenewal()` helper?

File: pkg/crc/cluster/cert-renewal.go
Comment (cfergeau) [inline pkg/crc/cluster/cert-renewal.go:58]: This seems to be explaining why waitForPendingCSRs/ApproveNodeCSR was removed in the `if server` case. This should be mentioned/explained in the commit log as well.

**Review by cfergeau — COMMENTED** [review]
>  Move RegenerateCertificates right after kubelet startup in machine.Start 

> cluster.RegenerateCertificates doesn't need to stop/start kubelet anymore.

Maybe add explicitly that it's not needed anymore with newer openshift releases, since kubelet uses the new certificate without restarting?


> Verify both client and server kubelet certificates

> When approving CSR, now we select exactly the one we expect by specifying the signerName.

It would be good for future reference to expand this commit log to describe more clearly the behaviour before, the new behaviour, ..
Before we were only checking the client certificate `/var/lib/kubelet/pki/kubelet-client-current.pem`, but expecting 2 CSRs, one which is generated when the client certificate is expired, and another one which is generated when the server certificate is expired. After this PR, crc also checks for expiration of the server certificate ``/var/lib/kubelet/pki/kubelet-server-current.pem`, and `WaitAndApprovePendingCSRs` has been improved so that depending on which certificate(s) has expired, it waits/approves the correct CSR.


It might also be worth adding 
> Yes the serving cert is an extra check. With current master, it is silently updated (during the 3min sleep time maybe). Until the cert is updated, user can't exec in a container nor fetching container logs.
I hope to remove the final sleep(3 min) with this kind of checks.

in one of the commits logs. The more we record regarding why a change is being made, our intent, ..., the easier it will be if we look at this code again a few months/years from now after having forgotten most of these discussions.

File: pkg/crc/cluster/cert-renewal.go
Comment (cfergeau) [inline pkg/crc/cluster/cert-renewal.go:15]: Since you are changing the method, you can fix the spelling `waitForPendingCSRs` to be consistent with the other method with `CSRs` in its name.

File: pkg/crc/machine/machine.go
Comment (guillaumerose) [reply pkg/crc/machine/machine.go:339]: yes done!

File: pkg/crc/cluster/cert-renewal.go
Comment (guillaumerose) [reply pkg/crc/cluster/cert-renewal.go:38]: yes changed!

File: pkg/crc/cluster/cert-renewal.go
Comment (guillaumerose) [reply pkg/crc/cluster/cert-renewal.go:58]: yes changelog is better now

File: pkg/crc/cluster/cert-renewal.go
Comment (guillaumerose) [reply pkg/crc/cluster/cert-renewal.go:52]: yes done!

File: pkg/crc/cluster/cert-renewal.go
Comment (guillaumerose) [reply pkg/crc/cluster/cert-renewal.go:15]: done

**Review by cfergeau — APPROVED** [review]

**Pull request merged**
